_posts/2020-06-12-terraform-azurerm-azure-container-instance-devops-agent.md

```diff
@@ -0,0 +1,135 @@
+
+## When to use Azure DevOps self-hosted agents?
+
+In my opinion, there are only a few scenarios that justify to not use the Microsoft-managed agents:
+
+1. You must be inside a private virtual network, control the network traffic, block the Internet access from the agent etc.
+2. You are using softwares that are not available on the Microsoft-hosted agents, that cannot be installed at build time because they are too big or require GUI
+3. You need the state on the agent to be maintained between two executions / job / stages, so you need to be able to target a machine explicitly
+4. You require specific hardware (GPU, custom boards etc...)
+
+If you are not in one of these cases, it should be possible for you to use any of the [Windows, Linux or macOS Microsoft-managed agents](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml).
+
+There are different ways to create your own self-hosted agents:
+
+- it can be a physical or virtual machine running [Linux](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/v2-linux?view=azure-devops), [Windows](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/v2-windows?view=azure-devops) or [macOS](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/v2-osx?view=azure-devops) where you install and configure the Azure DevOps agent software
+- it can be a [Docker container](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops)
+- it can be a a set of virtual machines, living in an [Azure ScaleSet that you let Azure DevOps manage](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/scale-set-agents?view=azure-devops) (this feature is still in preview at the time I am writing this article)
+
+Using the [terraform-azurerm-aci-devops-agent module](https://github.com/Azure/terraform-azurerm-aci-devops-agent), the agents will be running inside Docker containers, Linux or Windows, hosted on Azure Container Instance.
+
+## Build your own Docker image to run the DevOps agents
+
+Before being able to run the module, you need to build your own Docker image, that contains the DevOps agents and any tools that you want to embed into the Docker containers and that you cannot install at build time. To do that, you can follow the [official documentation](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops) or start from [the Dockerfile](https://github.com/Azure/terraform-azurerm-aci-devops-agent/blob/master/Docker/README.md) in the Terraform module repository.
+
+Once your Docker image is ready, you need to push it into a registry. It could be the public Docker hub, the Azure Container Registry or any Docker private registry.
```

Comment:
```suggestion
Once the Docker image is ready, we need to push it into a registry. It could be the public Docker hub, the Azure Container Registry or any Docker private registry.
```